File: leetcode/problems/language-specific/go/59.spiral-matrix-ii.go
/*
 * @lc app=leetcode id=59 lang=golang
 *
 * [59] Spiral Matrix II
 */

// @lc code=start
func generateMatrix(n int) [][]int {
	res, element := make([][]int, n), 1
	left, right, top, bottom := 0, n-1, 0, n-1
	for i := 0; i < n; i++ {
		res[i] = make([]int, n)
	}

	for left <= right && top <= bottom {
		// * Traverse from top left corner <=> top right corner
		for j := left; j <= right; j++ {
			res[top][j] = element
			element++
		}

		// * Traverse from top right corner <=> bottom right corner
		for i := top + 1; i <= bottom; i++ {
			res[i][right] = element
			element++
		}

		// * Traverse from bottom right corner <=> bottom left corner
		for j := right - 1; j >= left; j-- {
			res[bottom][j] = element
			element++
		}

		// * Traverse from bottom left corner <=> top left corner
		for i := bottom - 1; i > top; i-- {
			res[i][left] = element
			element++
		}

		left++
		right--
		top++
		bottom--
	}

	return res
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-09-16/59.spiral-matrix-ii.py
#
# @lc app=leetcode id=59 lang=python3
#
# [59] Spiral Matrix II
#

# @lc code=start

# * Iterative Solution | O(mn) Time | O(1) Space
# * m -> The number of rows in the matrix | n -> The number of columns in the matrix


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        left, right = 0, n
        top, bottom = 0, n
        count = 1
        matrix = [[0 for _ in range(n)] for _ in range(n)]
        while left < right and top < bottom:
            for col in range(left, right):
                matrix[top][col] = count
                count += 1
            top += 1

            for row in range(top, bottom):
                matrix[row][right - 1] = count
                count += 1
            right -= 1

            if not left < right or not top < bottom:
                break

            for col in range(right - 1, left - 1, -1):
                matrix[bottom - 1][col] = count
                count += 1
            bottom -= 1

            for row in range(bottom - 1, top - 1, -1):
                matrix[row][left] = count
                count += 1
            left += 1

        return matrix


# @lc code=end
